src/core/tUnification.ml

```diff
@@ -968,37 +976,54 @@ and unifies_to_field uctx a b ab tl (t,cf) =
+		| t -> t
+	in
+	let compare_underlying equality_kind a b = type_eq {uctx with equality_kind = equality_kind} a b in
+	let unifies_abstract uctx t ab tl ats =
+		List.exists (does_func_unify_arg (fun at ->
+			let at = apply_params ab.a_params tl at in
+			if ats == ab.a_to then
+				with_variance uctx f at t
+			else
+				with_variance uctx f t at
+		)) ats
+	in
+	let fail () = error [cannot_unify t1 t2] in
+	match t1,t2 with
 	| TInst(c1,tl1),TInst(c2,tl2) when c1 == c2 ->
-		List.iter2 f tl1 tl2
+		List.iter2 unify_param tl1 tl2
 	| TEnum(en1,tl1),TEnum(en2,tl2) when en1 == en2 ->
-		List.iter2 f tl1 tl2
-	| TAbstract(a1,tl1),TAbstract(a2,tl2) when a1 == a2 && Meta.has Meta.CoreType a1.a_meta ->
-		List.iter2 f tl1 tl2
-	| TAbstract(a1,pl1),TAbstract(a2,pl2) ->
-		if (Meta.has Meta.CoreType a1.a_meta) && (Meta.has Meta.CoreType a2.a_meta) then begin
-			let ta1 = apply_params a1.a_params pl1 a1.a_this in
-			let ta2 = apply_params a2.a_params pl2 a2.a_this in
-			type_eq {uctx with equality_kind = EqStrict} ta1 ta2;
-		end;
-		if not (List.exists (allows_variance_to t2) a1.a_to) && not (List.exists (allows_variance_to t1) a2.a_from) then
-			error [cannot_unify t1 t2]
-	| TAbstract(a,pl),t ->
-		type_eq {uctx with equality_kind = EqBothDynamic} (apply_params a.a_params pl a.a_this) t;
-		if not (List.exists (fun t2 -> allows_variance_to t (apply_params a.a_params pl t2)) a.a_to) then error [cannot_unify t1 t2]
-	| t,TAbstract(a,pl) ->
-		type_eq {uctx with equality_kind = EqBothDynamic} t (apply_params a.a_params pl a.a_this);
-		if not (List.exists (fun t2 -> allows_variance_to t (apply_params a.a_params pl t2)) a.a_from) then error [cannot_unify t1 t2]
-	| (TAnon a1 as t1), (TAnon a2 as t2) ->
-		rec_stack unify_stack (t1,t2)
-			(fun (a,b) -> fast_eq a t1 && fast_eq b t2)
-			(fun() -> unify_anons uctx t1 t2 a1 a2)
-			(fun l -> error l)
+		List.iter2 unify_param tl1 tl2
+	| TAbstract(a1,tl1),TAbstract(a2,tl2) when a1 == a2 ->
+		List.iter2 unify_param tl1 tl2
+	| TAbstract(a1,tl1),TAbstract(a2,tl2) ->
+		let uctx = get_after_abstract_context() in
+		compare_underlying EqStrict (get_underlying_type t1) (get_underlying_type t2);
+		if not (unifies_abstract uctx t2 a1 tl1 a1.a_to) && not (unifies_abstract uctx t1 a2 tl2 a2.a_from) then fail();
+	| TAbstract(a,tl),_ ->
+		let uctx = get_after_abstract_context() in
+		compare_underlying EqBothDynamic (get_underlying_type t1) t2;
+		if not (unifies_abstract uctx t2 a tl a.a_to) then fail();
+	| _,TAbstract(a,tl) ->
+		let uctx = get_after_abstract_context() in
+		compare_underlying EqBothDynamic t1 (get_underlying_type t2);
+		if not (unifies_abstract uctx t1 a tl a.a_from) then fail()
+	| TAnon(a1),TAnon(a2) ->
+		rec_stack_default unify_stack (t1,t2) (fast_eq_pair (t1,t2)) (fun() -> unify_anons uctx t1 t2 a1 a2) ()
 	| _ ->
-		error [cannot_unify t1 t2]
+		fail()
 
 and unify_type_params uctx a b tl1 tl2 =
+	let uctx = get_nested_context uctx in
```

Comment:
Technically, this uctx should not go into `type_eq`, but I don't think it makes a difference in the current implementation.

Reply:
Question is why `type_eq` needs a unification context? I mean there is no need for it right now, but is there some idea for future property in the unification context that will be used both by `unify` and `type_eq`?

Reply:
I was contemplating merging `type_eq` and `unify`, but I've moved away from that idea. It's hard to tell if a unification context could be useful for it, so I wouldn't mind removing it again for now.